ql/instruments/creditdefaultswap.cpp

```diff
@@ -71,26 +97,56 @@ namespace QuantLib {
                                          bool paysAtDefaultTime,
                                          const Date& protectionStart,
                                          const Date& upfrontDate,
-                                         const boost::shared_ptr<Claim>& claim)
+                                         const boost::shared_ptr<Claim>& claim,
+                                         const DayCounter& lastPeriodDayCounter,
+                                         const bool rebatesAccrual)
     : side_(side), notional_(notional), upfront_(upfront),
       runningSpread_(runningSpread), settlesAccrual_(settlesAccrual),
       paysAtDefaultTime_(paysAtDefaultTime), claim_(claim),
       protectionStart_(protectionStart == Null<Date>() ? schedule[0] :
                                                          protectionStart) {
-        QL_REQUIRE(protectionStart_ <= schedule[0],
+        QL_REQUIRE((protectionStart_ <= schedule[0])
+            || (schedule.rule() == DateGeneration::CDS),
                    "protection can not start after accrual");
+
         leg_ = FixedRateLeg(schedule)
             .withNotionals(notional)
             .withCouponRates(runningSpread, dayCounter)
-            .withPaymentAdjustment(convention);
-        Date d = upfrontDate == Null<Date>() ? schedule[0] : upfrontDate;
-        upfrontPayment_.reset(new SimpleCashFlow(notional*upfront, d));
-        QL_REQUIRE(upfrontPayment_->date() >= protectionStart_,
+            .withPaymentAdjustment(convention)
+            .withLastPeriodDayCounter(lastPeriodDayCounter);
+
+        // If empty, adjust to T+3 standard settlement, alternatively add
+        //  an arbitrary date to the constructor
+        Date effectiveUpfrontDate = upfrontDate == Null<Date>() ?
+            schedule.calendar().advance(
+            schedule.calendar().adjust(protectionStart_, convention),
+            2, Days, convention) : upfrontDate;
```

Comment:
When is the above different from `schedule.calendar().advance(protectionStart_, 2, Days, convention)`? As far as I can see, the initial adjustment should be taken care of.